src/scripts/gap_detector.py

```diff
@@ -0,0 +1,147 @@
+                    params=[
+                        QueryParameter.number_type("start", start),
+                        QueryParameter.number_type("end", end),
+                    ],
+                )
+            ).data
+        )
+
+    def db_df(self, query_str: str, start: int, end: int) -> DataFrame:
+        """Executes and fetches dataframe for `query_str` from Database"""
+        return self.database.exec_query(
+            query_str.replace("{{start}}", str(start)).replace("{{end}}", str(end)),
+            self.database.connections[0],
+        )
+
+    def tx_list_diff(
+        self,
+        start: int,
+        end: int,
+    ) -> set[str]:
+        """Gets the list of differing transaction hashes between Dune and Database"""
+        dune_hashes = set(self.dune_df(DUNE_HASH_QUERY_ID, start, end)["tx_hash"])
+        db_hashes = set(self.db_df(DB_HASH_QUERY, start, end)["tx_hash"])
+        # TODO - usually its the database missing something (i.e. we assume Dune contains all).
+        #  However, we could also return an object that keeps track of what is missing where.
+        return dune_hashes.symmetric_difference(db_hashes)
+
+    def count_diff(
+        self,
+        start: int,
+        end: int,
+    ) -> int:
+        """
+        This function acts as a quick comparison between data sources for a block range.
+        The assumption is that if the number of batches between a certain block range agrees,
+        between the two sources then they are correct. If a difference is detected, one could
+        drill deeper and compare the set of transaction hashes. The transaction hash comparison
+        would also reveal any other types of disagreement that the count does not catch.
+        """
+        dune_count = self.dune_df(DUNE_COUNT_QUERY_ID, start, end)
+        db_count = self.db_df(DB_COUNT_QUERY, start, end)
+
+        return int(dune_count["batches"][0]) - int(db_count["batches"][0])
+
+    def find_missing(
+        self,
+        start: int,
+        end: int,
+    ) -> set[str]:
+        """
+        A recursive binary search, returning any disagreement
+        in transaction hash sets (between dune and database) for a given block range.
+        """
+        print("Inspecting Block Range...", start, end)
+        count_diff = self.count_diff(start, end)
+        if count_diff == 0:
+            # Nothing missing here!
+            print("counts agree on", start, end)
+            return set()
+
+        if count_diff < 200:  # TODO - make this configurable.
+            # getting down into the trees.
+            diff = self.tx_list_diff(start, end)
+            print("Diff", diff)
+            if diff:
+                return diff
+
+        print("count diff", count_diff)
+        mid = (start + end) // 2
+        # Note that this implementation always prefers left side first.
+        return self.find_missing(start, mid).union(self.find_missing(mid + 1, end))
```

Comment:
Out of curiosity: What was the reasoning behind using a binary search instead of searching the whole set at once?

Reply:
Its about optimizing the number of datapoints being from the Dune Query Results. Also, when I was looking into this manually it was hard for me to compare a list of 100 thousand transaction hashes to see which one was missing. Dune definitely charges for data points (cf [api docs pricing](https://dune.com/docs/api/#api-pricing)), so I wanted to reduce this. 

However, reducing this by count_diff is not really the right approach. What we really want to do is to only fetch hashes when the block range is small enough. Maybe we should change the condition above to be `if block_range is less than 1000 blocks` which is about 3.5 hours.

Reply:
Thanks for elaborating on this. This PR makes a lot more sense now.

Just thinking out loud here: Is there a max character limit on queries in Dune? Theoretically, could we just upload all our DB tx_hash-es into a query and say, `select * from dune_tx_hashes where tx_hash not in (our hashes)`? This way, we would the calculation inside of Dune's SQL and save money without having to fetch data points.

Let me know what you think about this.

Reply:
I think that is a cool idea too and did definitely consider it. It just seems like the code is a bit lopsided/anti-symmetric if we take that approach. Its also a lot of information being passed around. Somehow I feel like the least bandwidth solution is the most elegant.

Reply:
Good to know you've considered it.